File: scripts/export_bigwig_chrom_bedgraph.R
#!/usr/bin/env Rscript

usage <- function(status = 0L) {
    stream <- if (status == 0L) stdout() else stderr()
    writeLines(c(
        "Usage: export_bigwig_chrom_bedgraph.R --input FILE.bigWig",
        "       --output FILE.bedGraph [--chrom CHROM]",
        "",
        "Export one chromosome from a BigWig as sorted, non-overlapping bedGraph.",
        "Only finite signal values greater than zero are retained. Coordinates in",
        "the bedGraph output are BED 0-based half-open coordinates.",
        "",
        "Options:",
        "  --input FILE    Source BigWig",
        "  --output FILE   New bedGraph; an existing file is never replaced",
        "  --chrom CHROM   Chromosome with or without chr prefix (default: 1)",
        "  -h, --help      Show this help"
    ), con = stream)
    quit(status = status)
}

arguments <- commandArgs(trailingOnly = TRUE)
if (any(arguments %in% c("-h", "--help"))) {
    usage()
}

values <- list(chrom = "1")
index <- 1L
while (index <= length(arguments)) {
    option <- arguments[[index]]
    if (!option %in% c("--input", "--output", "--chrom")) {
        writeLines(paste("E: unknown argument:", option), con = stderr())
        usage(2L)
    }
    index <- index + 1L
    if (index > length(arguments)) {
        writeLines(paste("E:", option, "requires a value"), con = stderr())
        usage(2L)
    }
    values[[substring(option, 3L)]] <- arguments[[index]]
    index <- index + 1L
}

if (is.null(values$input) || is.null(values$output)) {
    writeLines("E: --input and --output are required", con = stderr())
    usage(2L)
}
if (!file.exists(values$input)) {
    stop("BigWig input does not exist: ", values$input, call. = FALSE)
}
if (file.exists(values$output)) {
    stop("refusing to replace existing output: ", values$output, call. = FALSE)
}

suppressPackageStartupMessages({
    library(GenomicRanges)
    library(rtracklayer)
})

normalize_chrom <- function(chrom) sub("^chr", "", chrom, ignore.case = TRUE)

bigwig <- BigWigFile(values$input)
available <- seqlevels(seqinfo(bigwig))
matches <- available[normalize_chrom(available) == normalize_chrom(values$chrom)]
if (length(matches) != 1L) {
    stop(
        "BigWig does not contain exactly one chromosome matching ",
        values$chrom,
        call. = FALSE
    )
}

chrom <- matches[[1L]]
chrom_length <- seqlengths(seqinfo(bigwig))[[chrom]]
if (is.na(chrom_length) || chrom_length <= 0) {
    stop("BigWig chromosome has no usable length: ", chrom, call. = FALSE)
}

ranges <- import(
    bigwig,
    which = GRanges(chrom, IRanges(start = 1L, end = chrom_length))
)
scores <- mcols(ranges)$score
ranges <- ranges[is.finite(scores) & scores > 0]

dir.create(dirname(values$output), recursive = TRUE, showWarnings = FALSE)
export(ranges, values$output, format = "bedGraph")
message(
    "I: exported ", length(ranges), " positive ", chrom,
    " intervals to ", values$output
)

File: scripts/export_bigwig_chrom_bedgraph.py
#!/usr/bin/env python3

"""Export one BigWig chromosome as positive, non-overlapping bedGraph rows."""

from __future__ import annotations

import argparse
import math
import os
import sys
from pathlib import Path


def normalize_chrom(value: str) -> str:
    normalized = value[3:] if value.lower().startswith("chr") else value
    return "MT" if normalized.upper() in {"M", "MT"} else normalized


def chrom_matches(observed: str, requested: str,
                  mitochondrial_aliases: bool) -> bool:
    observed_normalized = normalize_chrom(observed).upper()
    requested_normalized = normalize_chrom(requested).upper()
    if mitochondrial_aliases:
        aliases = {"M", "MT", "25"}
        if requested_normalized not in aliases:
            raise RuntimeError(
                "--mitochondrial-aliases requires --chrom M, MT, or 25"
            )
        return observed_normalized in aliases
    return observed_normalized == requested_normalized


def export(arguments: argparse.Namespace) -> None:
    try:
        import pyBigWig  # type: ignore[import-not-found]
    except ImportError as error:
        raise RuntimeError(
            "pyBigWig is required; install the conda-forge pybigwig package"
        ) from error

    source = arguments.input.expanduser().resolve()
    output = arguments.output.expanduser().resolve()
    if not source.is_file():
        raise RuntimeError(f"BigWig input not found: {source}")
    if output.exists():
        raise RuntimeError(f"refusing to replace existing output: {output}")
    output.parent.mkdir(parents=True, exist_ok=True)

    with pyBigWig.open(str(source)) as bigwig:
        matching = [
            chrom for chrom in bigwig.chroms()
            if chrom_matches(
                chrom, arguments.chrom, arguments.mitochondrial_aliases
            )
        ]
        if not matching and arguments.allow_empty:
            chrom = arguments.chrom
            intervals = ()
        elif len(matching) != 1:
            raise RuntimeError(
                f"BigWig has {len(matching)} chromosomes matching {arguments.chrom}"
            )
        else:
            chrom = matching[0]
            intervals = bigwig.intervals(chrom) or ()
        temporary = output.with_name(f".{output.name}.tmp-{os.getpid()}")
        written = 0
        try:
            with temporary.open("w", encoding="ascii", newline="") as handle:
                for start, end, value in intervals:
                    if math.isfinite(value) and value > 0:
                        handle.write(f"{chrom}\t{start}\t{end}\t{value:.17g}\n")
                        written += 1
            if written == 0 and not arguments.allow_empty:
                raise RuntimeError(
                    f"BigWig has no positive intervals on chromosome {chrom}"
                )
            os.replace(temporary, output)
        finally:
            if temporary.exists():
                temporary.unlink()
    print(
        f"I: exported {written} positive {chrom} intervals: {output}",
        file=sys.stderr,
    )


def parser() -> argparse.ArgumentParser:
    result = argparse.ArgumentParser(
        description=(
            "Export one chromosome from BigWig to sorted positive-depth "
            "bedGraph using 0-based half-open coordinates."
        )
    )
    result.add_argument("--input", type=Path, required=True)
    result.add_argument("--output", type=Path, required=True)
    result.add_argument("--chrom", default="1")
    result.add_argument(
        "--allow-empty", action="store_true",
        help=(
            "write a valid empty bedGraph when the chromosome has no positive "
            "intervals; useful for whole-genome zero-support strata"
        ),
    )
    result.add_argument(
        "--mitochondrial-aliases", action="store_true",
        help=(
            "for a run-plan-identified mitochondrial sequence, treat M, MT, "
            "and 25 (with optional chr prefix) as equivalent"
        ),
    )
    return result


def main() -> int:
    arguments = parser().parse_args()
    try:
        export(arguments)
    except (OSError, RuntimeError, ValueError) as error:
        print(f"E: {error}", file=sys.stderr)
        return 1
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
